src/libcore/ffi.rs

```diff
@@ -53,80 +52,158 @@ impl fmt::Debug for c_void {
+          not(windows)))]
 #[unstable(feature = "c_variadic",
            reason = "the `c_variadic` feature has not been properly tested on \
                      all supported platforms",
            issue = "44930")]
-struct VaListImpl {
-    gp_offset: i32,
-    fp_offset: i32,
-    overflow_arg_area: *mut c_void,
-    reg_save_area: *mut c_void,
+mod impls {
+    use crate::marker::PhantomData;
+    use crate::ops::{Deref, DerefMut};
+    use super::c_void;
+
+    /// AArch64 ABI implementation of a `va_list`. See the
+    /// [AArch64 Procedure Call Standard] for more details.
+    ///
+    /// [AArch64 Procedure Call Standard]:
+    /// http://infocenter.arm.com/help/topic/com.arm.doc.ihi0055b/IHI0055B_aapcs64.pdf
+    #[cfg(all(target_arch = "aarch64", not(target_os = "ios"), not(windows)))]
+    #[repr(C)]
+    #[derive(Debug)]
+    #[lang = "va_list"]
+    pub struct VaListImpl<'f> {
+        stack: *mut c_void,
+        gr_top: *mut c_void,
+        vr_top: *mut c_void,
+        gr_offs: i32,
+        vr_offs: i32,
+        _marker: PhantomData<&'f c_void>,
+    }
+
+    /// PowerPC ABI implementation of a `va_list`.
+    #[cfg(all(target_arch = "powerpc", not(windows)))]
+    #[repr(C)]
+    #[derive(Debug)]
+    #[lang = "va_list"]
+    pub struct VaListImpl<'f> {
+        gpr: u8,
+        fpr: u8,
+        reserved: u16,
+        overflow_arg_area: *mut c_void,
+        reg_save_area: *mut c_void,
+        _marker: PhantomData<&'f c_void>,
+    }
+
+    /// x86_64 ABI implementation of a `va_list`.
+    #[cfg(all(target_arch = "x86_64", not(windows)))]
+    #[repr(C)]
+    #[derive(Debug)]
+    #[lang = "va_list"]
+    pub struct VaListImpl<'f> {
+        gp_offset: i32,
+        fp_offset: i32,
+        overflow_arg_area: *mut c_void,
+        reg_save_area: *mut c_void,
+        _marker: PhantomData<&'f c_void>,
+    }
+
+    impl<'f> VaListImpl<'f> {
+        /// Convert a `VaListImpl` into a `VaList` that is binary-compatible with C's `va_list`.
+        #[inline]
+        pub fn as_va_list<'s>(&'s mut self) -> VaList<'s, 'f> {
+            VaList(self)
+        }
+    }
+
+    /// A wrapper for a `va_list`
+    #[repr(transparent)]
+    #[derive(Debug)]
+    pub struct VaList<'a, 'f: 'a>(&'a mut VaListImpl<'f>);
+
+    impl<'a, 'f: 'a> Deref for VaList<'a, 'f> {
+        type Target = VaListImpl<'f>;
+
+        #[inline]
+        fn deref(&self) -> &VaListImpl<'f> {
+            self.0
+        }
+    }
+
+    impl<'a, 'f: 'a> DerefMut for VaList<'a, 'f> {
+        #[inline]
+        fn deref_mut(&mut self) -> &mut VaListImpl<'f> {
+            self.0
+        }
+    }
 }
 
-/// A wrapper for a `va_list`
-#[lang = "va_list"]
-#[derive(Debug)]
 #[unstable(feature = "c_variadic",
            reason = "the `c_variadic` feature has not been properly tested on \
                      all supported platforms",
            issue = "44930")]
-#[repr(transparent)]
-pub struct VaList<'a>(&'a mut VaListImpl);
+pub use self::impls::*;
```

Comment:
Can you replace `*` with `VaListImpl`? That might fix the paths reported elsewhere to be `std::ffi::VaListImpl` instead of `core::ffi::impls::VaListImpl` (explicit reexports are needed to give a "shorter path").

Also, can you leave a comment about the final name still being undecided?

Reply:
I changed this to `::{VaList, VaListImpl}`, and it didn't fix the paths in the errors.

Reply:
Huh, weird, I wonder if it's because the module they're defined in is private, but maybe it's some other subtlety regarding reexports (don't we have a lot of items in `std::iter` that are defined in private submodules of `core::iter`?).
cc @alexcrichton